### PR TITLE
Fix API `version` and `upgrade_version` filters to work with different Wazuh version formats

### LINKED ISSUE #10426: Investigate to make the API `version` filter work with different Wazuh version formats
|Wazuh version|Component|
|---|---|
| 4.4.0 | API |

Right now, the `version` parameter used to filter by agent version in the API requests only works if we specify the agent version following the format: `Wazuh vX.Y.Z`

https://github.com/wazuh/wazuh/blob/9d109cc7e88c636cf70b911ad3e7ff4042eca86c/api/api/spec/spec.yaml#L4566-L4572

In this issue, we should investigate how to make the filter work with other Wazuh version formats. Formats like `X.Y.Z` or `vX.Y.Z` should be allowed when using this filter.

In case we cannot add these new formats, we could create a new validator in `validator.py` to indicate the specific format that the `version` parameter must follow.

#### Example:

Request

`GET /agents?select=version&version=Wazuh%20v4.3.0
`

Response: 

```
{
  "data": {
    "affected_items": [
      {
        "version": "Wazuh v4.3.0",
        "id": "000"
      }
    ],
    "total_affected_items": 1,
    "total_failed_items": 0,
    "failed_items": []
  },
  "message": "All selected agents information was returned",
  "error": 0
}
```

Request

`GET /agents?select=version&version=v4.3.0
`

Response

```
{
  "data": {
    "affected_items": [],
    "total_affected_items": 0,
    "total_failed_items": 0,
    "failed_items": []
  },
  "message": "No agent information was returned",
  "error": 0
}
```

### wazuh/wazuh
- Unit tests without failures. Updated and/or expanded if there are new functions/methods/outputs:
  - [x] Cluster (`framework/wazuh/core/cluster/tests/` & `framework/wazuh/core/cluster/dapi/tests/`)
  - [x] Core (`framework/wazuh/core/tests/`)
  - [x] SDK (`framework/wazuh/tests/`)
  - [x] RBAC (`framework/wazuh/rbac/tests/`)
  - [x] API (`api/api/tests/`)
- API tavern integration tests without failures. Updated and/or expanded if needed (`api/test/integration/`):
  - [x] Affected tests 
  - [x] Affected RBAC (black and white) tests
- [x] Review integration test mapping using the script (`api/test/integration/mapping/integration_test_api_endpoints.json`)
- [x] Review of spec.yaml examples and schemas (`api/api/spec/spec.yaml`)
- [x] Review exceptions remediation when any endpoint path changes or is removed (`framework/wazuh/core/exception.py`)
- [x] Changelog (`CHANGELOG.md`)

### wazuh/wazuh-documentation
- [x] Migration from 3.X for changed endpoints (`source/user-manual/api/equivalence.rst`)
- [x] Update RBAC reference with new/modified actions/resources/relationships (`source/user-manual/api/rbac/reference.rst`)

### PR DESCRIPTION
**Pull request by Kondent**
|Related issue|
|---|
|closes #10426  |

# Description
Under this development the format of the parameter `version` was unified to the same format that is stored into the database.

From now on, these four queries will be valid:
* `GET /agents?limit=1&select=version&version=v4.4.0`.
* `GET /agents?limit=1&select=version&version=4.4.0`.
* `GET /agents?limit=1&select=version&version=wazuh%204.4.0`.
* `GET /agents?limit=1&select=version&version=WaZuH%20V4.4.0`.

* Response:
```json
{
  "data": {
    "affected_items": [
      {
        "version": "Wazuh v4.4.0",
        "id": "000"
      }
    ],
    "total_affected_items": 6,
    "total_failed_items": 0,
    "failed_items": []
  },
  "message": "All selected agents information was returned",
  "error": 0
}
```

With any other different value from the four mentioned above:
* Response:
```json
{
  "data": {
    "affected_items": [],
    "total_affected_items": 0,
    "total_failed_items": 0,
    "failed_items": []
  },
  "message": "No agent information was returned",
  "error": 0
}
```

Plus, the validator used for the `upgrade_version` parameter, was updated as well.

### `version`
The verification and formatting is currently being done at `WazuhDBQueryAgents`.
This way, the parameter is formatted before preparing the database query.
_Format: `wazuh vX.Y.Z`_.
_NOTE: this modification could be done as a decorator too. However, since there are several implicated endpoints, I think this is a better way._

### `upgrade_version`:
The formatting is currently being done at `core_upgrade_agents @ core/agent.py`.
_Format: `vX.Y.Z`_.

# Unit test - API
```s
⩘ ~/git/wazuh/framework ⩫ fix/10426-wazuh-version-format ⨍ cd ~/git/wazuh/api && python3 -m pytest -x api --disable-warnings
==================================================================================== test session starts =====================================================================================
platform linux -- Python 3.9.5, pytest-5.4.3, py-1.11.0, pluggy-0.13.1
rootdir: /home/kondent/git/wazuh/api
plugins: html-2.1.1, asyncio-0.15.1, metadata-1.11.0, aiohttp-0.3.0, trio-0.7.0, cov-2.10.1
collected 509 items                                                                                                                                                                          

api/controllers/test/test_active_response_controller.py .                                                                                                                              [  0%]
api/controllers/test/test_agent_controller.py ..........................................                                                                                               [  8%]
api/controllers/test/test_cdb_list_controller.py ......                                                                                                                                [  9%]
api/controllers/test/test_ciscat_controller.py .                                                                                                                                       [  9%]
api/controllers/test/test_cluster_controller.py ......................                                                                                                                 [ 14%]
api/controllers/test/test_decoder_controller.py .......                                                                                                                                [ 15%]
api/controllers/test/test_default_controller.py .                                                                                                                                      [ 15%]
api/controllers/test/test_experimental_controller.py ...............                                                                                                                   [ 18%]
api/controllers/test/test_manager_controller.py .................                                                                                                                      [ 22%]
api/controllers/test/test_mitre_controller.py .......                                                                                                                                  [ 23%]
api/controllers/test/test_overview_controller.py .                                                                                                                                     [ 23%]
api/controllers/test/test_rootcheck_controller.py ....                                                                                                                                 [ 24%]
api/controllers/test/test_rule_controller.py ........                                                                                                                                  [ 25%]
api/controllers/test/test_sca_controller.py ..                                                                                                                                         [ 26%]
api/controllers/test/test_security_controller.py ...................................................                                                                                   [ 36%]
api/controllers/test/test_syscheck_controller.py ....                                                                                                                                  [ 37%]
api/controllers/test/test_syscollector_controller.py .........                                                                                                                         [ 38%]
api/controllers/test/test_task_controller.py .                                                                                                                                         [ 39%]
api/controllers/test/test_vulnerability_controller.py ..                                                                                                                               [ 39%]
api/models/test/test_model.py ...........................                                                                                                                              [ 44%]
api/test/test_alogging.py ..........                                                                                                                                                   [ 46%]
api/test/test_authentication.py ..........                                                                                                                                             [ 48%]
api/test/test_configuration.py .........................................                                                                                                               [ 56%]
api/test/test_encoder.py ...                                                                                                                                                           [ 57%]
api/test/test_middlewares.py ............                                                                                                                                              [ 59%]
api/test/test_uri_parser.py ...                                                                                                                                                        [ 60%]
api/test/test_util.py ........................................                                                                                                                         [ 68%]
api/test/test_validator.py ........................................................................................................................................................... [ 98%]
.......                                                                                                                                                                                [100%]

============================================================================== 509 passed, 13 warnings in 2.07s ==============================================================================
```

# Unit test - framework
```s
⩘ ~/git/wazuh/api ⩫ fix/10426-wazuh-version-format ⨍ cd ~/git/wazuh/framework && python3 -m pytest -x wazuh --disable-warnings
==================================================================================== test session starts =====================================================================================
platform linux -- Python 3.9.5, pytest-5.4.3, py-1.11.0, pluggy-0.13.1
rootdir: /home/kondent/git/wazuh/framework
plugins: html-2.1.1, asyncio-0.15.1, metadata-1.11.0, aiohttp-0.3.0, trio-0.7.0, cov-2.10.1
collected 1853 items                                                                                                                                                                         

wazuh/core/cluster/dapi/tests/test_dapi.py ..............................                                                                                                              [  1%]
wazuh/core/cluster/tests/test_client.py ................                                                                                                                               [  2%]
wazuh/core/cluster/tests/test_cluster.py ..................................                                                                                                            [  4%]
wazuh/core/cluster/tests/test_common.py ..............................................................                                                                                 [  7%]
wazuh/core/cluster/tests/test_control.py .....                                                                                                                                         [  7%]
wazuh/core/cluster/tests/test_local_client.py .............                                                                                                                            [  8%]
wazuh/core/cluster/tests/test_local_server.py .......................                                                                                                                  [  9%]
wazuh/core/cluster/tests/test_master.py ..........................................                                                                                                     [ 12%]
wazuh/core/cluster/tests/test_server.py ...................                                                                                                                            [ 13%]
wazuh/core/cluster/tests/test_utils.py .......                                                                                                                                         [ 13%]
wazuh/core/cluster/tests/test_worker.py ...............................                                                                                                                [ 15%]
wazuh/core/tests/test_active_response.py ..............                                                                                                                                [ 15%]
wazuh/core/tests/test_agent.py ......................................................................................................................................................  [ 24%]
wazuh/core/tests/test_cdb_list.py ......................................                                                                                                               [ 26%]
wazuh/core/tests/test_common.py .......                                                                                                                                                [ 26%]
wazuh/core/tests/test_configuration.py ....................................                                                                                                            [ 28%]
wazuh/core/tests/test_database.py .............                                                                                                                                        [ 29%]
wazuh/core/tests/test_decoder.py ................                                                                                                                                      [ 30%]
wazuh/core/tests/test_exception.py ...                                                                                                                                                 [ 30%]
wazuh/core/tests/test_input_validator.py ...                                                                                                                                           [ 30%]
wazuh/core/tests/test_logtest.py ..                                                                                                                                                    [ 30%]
wazuh/core/tests/test_manager.py ...............                                                                                                                                       [ 31%]
wazuh/core/tests/test_mitre.py .............                                                                                                                                           [ 31%]
wazuh/core/tests/test_pyDaemonModule.py .....                                                                                                                                          [ 32%]
wazuh/core/tests/test_results.py ........................................                                                                                                              [ 34%]
wazuh/core/tests/test_rootcheck.py .............                                                                                                                                       [ 35%]
wazuh/core/tests/test_rule.py ......................                                                                                                                                   [ 36%]
wazuh/core/tests/test_sca.py ............                                                                                                                                              [ 36%]
wazuh/core/tests/test_security.py ............                                                                                                                                         [ 37%]
wazuh/core/tests/test_stats.py ............                                                                                                                                            [ 38%]
wazuh/core/tests/test_syscheck.py .......                                                                                                                                              [ 38%]
wazuh/core/tests/test_syscollector.py ...                                                                                                                                              [ 38%]
wazuh/core/tests/test_task.py ........                                                                                                                                                 [ 39%]
wazuh/core/tests/test_utils.py ....................................................................................................................................................... [ 47%]
.........................................................................                                                                                                              [ 51%]
wazuh/core/tests/test_vulnerability.py .                                                                                                                                               [ 51%]
wazuh/core/tests/test_wazuh_queue.py ....................                                                                                                                              [ 52%]
wazuh/core/tests/test_wazuh_socket.py ....................                                                                                                                             [ 53%]
wazuh/core/tests/test_wdb.py ........................                                                                                                                                  [ 54%]
wazuh/core/tests/test_wlogging.py ........                                                                                                                                             [ 55%]
wazuh/rbac/tests/test_auth_context.py ..                                                                                                                                               [ 55%]
wazuh/rbac/tests/test_decorators.py .........................................................................................................                                          [ 60%]
wazuh/rbac/tests/test_default_configuration.py .......................................................                                                                                 [ 63%]
wazuh/rbac/tests/test_orm.py ......................................................                                                                                                    [ 66%]
wazuh/rbac/tests/test_preprocessor.py ...........                                                                                                                                      [ 67%]
wazuh/tests/test_active_response.py ............                                                                                                                                       [ 68%]
wazuh/tests/test_agent.py .....................................................................................................................                                        [ 74%]
wazuh/tests/test_cdb_list.py .....................................................                                                                                                     [ 77%]
wazuh/tests/test_ciscat.py .................................                                                                                                                           [ 79%]
wazuh/tests/test_cluster.py .......                                                                                                                                                    [ 79%]
wazuh/tests/test_decoder.py ...................................                                                                                                                        [ 81%]
wazuh/tests/test_group.py ........                                                                                                                                                     [ 81%]
wazuh/tests/test_logtest.py ......                                                                                                                                                     [ 82%]
wazuh/tests/test_manager.py ....................................                                                                                                                       [ 84%]
wazuh/tests/test_mitre.py .......                                                                                                                                                      [ 84%]
wazuh/tests/test_rootcheck.py ..................................................                                                                                                       [ 87%]
wazuh/tests/test_rule.py ........................................................                                                                                                      [ 90%]
wazuh/tests/test_sca.py .......                                                                                                                                                        [ 90%]
wazuh/tests/test_security.py .........................................................................                                                                                 [ 94%]
wazuh/tests/test_stats.py .......                                                                                                                                                      [ 94%]
wazuh/tests/test_syscheck.py .........................                                                                                                                                 [ 96%]
wazuh/tests/test_syscollector.py ............                                                                                                                                          [ 96%]
wazuh/tests/test_task.py ............................                                                                                                                                  [ 98%]
wazuh/tests/test_vulnerability.py ...............................                                                                                                                      [100%]

====================================================================== 1853 passed, 3777 warnings in 284.22s (0:04:44) =======================================================================
```

### CODE REVIEW

**Review by mcarmona99 — CHANGES_REQUESTED** [review]
We talked about adapting `_wazuh_version` to be also used as a validator for the `version` filter (apart from `upgrade_version`) but we didn't take into account that this change would make requests like the following:

`GET /agents?version='wrong'`

give a spec error, whereas they used to give an empty response with status code 200.

In my opinion, this is not correct. We are restricting this parameter to be used with only 4 formats but in truth, the parameter could be used with any string. The issue was asking for the capability of using `vx.y.z`, `x.y.z` and `Wazuh x.y.z` apart from `Wazuh vx.y.z` but not restricting to these formats.

To sum up, I think that the `version` filter should only use the `unify_wazuh_version_format` and not the validator.

Apart from this, review API integration tests since I think there is a lack of new test cases.

**Review by mcarmona99 — APPROVED** [review]
LGTM

**Review by davidjiglesias — APPROVED** [review]
